### ISSUE TITLE
Plans for AT support?

### ISSUE THREAD

**Issue by JoackimPennerup** [embedpdf/embed-pdf-viewer#135]
I'm evaluating EmbedPDF for our needs and it looks like it fits most of our needs. However, one feature that is missing is output screen readers to read. 
Are there any plans for that? Otherwise I might consider making a PR for one. I'm contemplating a parallel html representation of the document built from the structure tags, that can be read by screen readers. Does that should like the right approach?

**Comment by bobsingor** [comment]
@JoackimPennerup Can you send an example PDF document that has structure tags? And how the output HTML should look like? I am interested in this.

**Comment by JoackimPennerup** [comment]
I have done some experiments (together with Codex as PoC) on https://github.com/JoackimPennerup/embed-pdf-viewer/tree/codex/add-plugin-for-structelement-extraction.

The basic idea is to extract the tree of structure elements from the PDF, and create a hidden HTML version in the dom. 
This is what Adobe Acrobat finds in the ebook example:
<img width="275" height="576" alt="Image" src="https://github.com/user-attachments/assets/a1774176-9aa2-4872-ac95-970ca97b0994" />
This is an example of how your ebook-pdf would look like in the DOM:
```html
<img src="blob:http://127.0.0.1:3000/0225207c-ef0f-4cd9-b8a7-e9b45f17fcaa" aria-hidden="true" style="width: 100%; height: 100%; pointer-events: none;">
<div class="pdf-embed-a11y-helper">
    <article lang="en" data-pdftag="Document">
        <span data-pdftag="NonStruct">
            <span data-pdftag="NonStruct">
                <div data-pdftag="Div">
                    <span data-pdftag="NonStruct" style="left: 321.63px; top: 188.75px; width: 394.11px; height: 57.38px;">EmbedPDF</span>
                </div>
                <div data-pdftag="Div" style="left: 0px; top: 0px; width: 0px; height: 0px;">
                    <span data-pdftag="NonStruct" style="left: 164.59px; top: 342.77px; width: 530.01px; height: 46.81px;">PDF Viewers Are a</span>
                    <span data-pdftag="NonStruct" style="left: 164.59px; top: 412.23px; width: 566.25px; height: 63.42px;">Pain But they Don’t</span>
                    <span data-pdftag="NonStruct" style="left: 164.59px; top: 487.73px; width: 323.14px; height: 55.87px;">Have to Be)</span>
                </div>
                <div data-pdftag="Div" style="left: 0px; top: 0px; width: 0px; height: 0px;">
                    <span data-pdftag="NonStruct" style="left: 164.59px; top: 591.92px; width: 510.38px; height: 40.77px;">How EmbedPDF is Making</span>
                    <span data-pdftag="NonStruct" style="left: 164.59px; top: 643.26px; width: 456.02px; height: 40.77px;">PDF Integration Actually</span>
                    <span data-pdftag="NonStruct" style="left: 164.59px; top: 694.6px; width: 181.2px; height: 40.77px;">Enjoyable</span>
                </div>
                <div data-pdftag="Div" style="left: 0px; top: 0px; width: 0px; height: 0px;">
                    <span data-pdftag="NonStruct" style="left: 163.08px; top: 788.22px; width: 234.05px; height: 45.3px;">Bob Singor</span>
                </div>
                <div data-pdftag="Div" style="left: 0px; top: 0px; width: 0px; height: 0px;">
                    <span data-pdftag="NonStruct" style="left: 625.14px; top: 966.4px; width: 84.56px; height: 33.22px;">PDF</span>
                </div>
            </span></span>
    </article>
</div>
```
The data-pdftag attribute is just added for debugging. The positioning is to allow AT to highlight the right area when moving the virtual cursor. 
In a well tagged PDF you would normally find things like a H1 - H4, P, Figure, TOC, and so on. 
My experiments still have some kinks left: 

- Alt-texts for images
- language attributes where needed and extracted the correct way 
- Inherited styling in a suitable way.

Have a look and tell me what you think.

**Comment by bobsingor** [comment]
@JoackimPennerup Wow this is awesome work that you did here. Very valuable and the community will love this. Whenever you are ready I am happy to receive a pull request for this and I will include it in the library.

**Comment by JoackimPennerup** [comment]
@bobsingor, I've done some more testing and found that there is an API missing in PDFium. I'm not able to get the language for the document (/lang), only the structure elements. Should I push for a upstream change or do you maintain a fork already? I know that some wrappers expose GetRoot, to allow you to get arbitrary elements from the catalog, but that method is not available in the official build.

**Comment by bobsingor** [comment]
<img width="463" height="474" alt="Image" src="https://github.com/user-attachments/assets/3203a984-db48-4a6c-808f-494b69d83685" />

@JoackimPennerup There are potentially 2 places where there is a language see the picture above. In the Root! Or in the first child of the /StructTreeRoot 

Type = /StructElem
S = /Document
Lang = "en"
P (dict) [id: 46, gen: 0]
K (array)

You see? So probably you want to be able to access both. So you can fallback on the Root Lang. 

None of these are accessible in PDFium? I am happy to add those functions to PDFium, just let me know if you are talking about the one that I mention above or if I am missing something here.

**Comment by JoackimPennerup** [comment]
@bobsingor, yes one of my test documents only had the root language set, and thus the document was read by the screen reader in the wrong language. StructTreeRoot language is already accessible from the current API (via FPDF_StructElement_GetLang).

**Comment by bobsingor** [comment]
@JoackimPennerup I just merged this: #150 which adds a function to get the root language

**Comment by bobsingor** [comment]
@JoackimPennerup How is this one going? Let me know if you need any help? Would love to receive a PR for this one!

**Comment by JoackimPennerup** [comment]
@bobsingor, I've high workload of other stuff right now. So I have not had the time to clean up the code I pair coded with Codex. But in a week or two things will slow down a bit and I will be able to put some hours into cleaning up the code for a proper PR.
